File: bot.js
import express from 'express';
import { Client, GatewayIntentBits, Partials } from 'discord.js';
import { joinVoiceChannel, createAudioPlayer, createAudioResource, AudioPlayerStatus } from '@discordjs/voice';
import fs from 'fs';
import path from 'path';
import dotenv from 'dotenv';
import { fileURLToPath } from 'url';

dotenv.config();

const app = express();
const PORT = process.env.PORT || 3000;

app.get('/', (req, res) => res.send('Bot is running!'));
app.listen(PORT, () => console.log(`Web server running on port ${PORT}`));

// Get the current directory in ES module mode
const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const audioFilePath = path.join(__dirname, 'clip1.mp3');

const client = new Client({
  intents: [
    GatewayIntentBits.Guilds,
    GatewayIntentBits.GuildMessages,
    GatewayIntentBits.MessageContent,
    GatewayIntentBits.GuildVoiceStates,
  ],
  partials: [Partials.Channel],
});

const donnieEnabled = new Map();

client.on('messageCreate', async (message) => {
  if (message.channel.name !== 'bot-commands' || message.author.bot) return;

  const args = message.content.trim().split(/\s+/);
  if (args[0] === '!donnie') {
    const setting = args[1]?.toLowerCase();
    if (setting === 'on') {
      donnieEnabled.set(message.guild.id, true);
      message.reply('ENABLED');
    } else if (setting === 'off') {
      donnieEnabled.set(message.guild.id, false);
      message.reply('DISABLED');
    } else {
      message.reply('Usage: `!donnie on` or `!donnie off`');
    }
  }
});

client.on('voiceStateUpdate', async (oldState, newState) => {
  const guildId = newState.guild.id;
  if (!donnieEnabled.get(guildId)) return;

  const member = newState.member;
  if (!member || member.user.bot || !newState.channel || oldState.channel === newState.channel) return;

  const connection = joinVoiceChannel({
    channelId: newState.channel.id,
    guildId: newState.guild.id,
    adapterCreator: newState.guild.voiceAdapterCreator,
    selfDeaf: false,
  });

  const player = createAudioPlayer();
  connection.subscribe(player);

  const receiver = connection.receiver;

  receiver.speaking.on('start', (userId) => {
    const user = newState.guild.members.cache.get(userId);
    if (user && !user.user.bot && player.state.status !== AudioPlayerStatus.Playing) {
      const resource = createAudioResource(audioFilePath);
      player.play(resource);
    }
  });

  // Leave after inactivity or disable
  const cleanupInterval = setInterval(() => {
    // Add a check to ensure `newState.channel` is valid
    if (!donnieEnabled.get(guildId) || !newState.channel || newState.channel.members.size <= 1) {
      connection.destroy();
      clearInterval(cleanupInterval);
    }
  }, 5000);
});


client.login(process.env.DCTOKEN);

File: bot.py
import discord
from discord.ext import commands
import asyncio
import os
DCTOKEN = os.environ.get('DCTOKEN')

intents = discord.Intents.default()
intents.message_content = True  # Add this
intents.voice_states = True
intents.members = True

bot = commands.Bot(command_prefix="!", intents=intents)

donnie_enabled = {}

@bot.command()
async def donnie(ctx, arg):
    if ctx.channel.name != "bot-commands":
        await ctx.send("Commands can only be used in #bot-commands.")
        return

    if arg.lower() == "on":
        donnie_enabled[ctx.guild.id] = True
        await ctx.send("ENABLED")
    elif arg.lower() == "off":
        donnie_enabled[ctx.guild.id] = False
        await ctx.send("DISABLED")

	
    else:
        await ctx.send("Usage: `!donnie on` or `!donnie off`")

@bot.event
async def on_voice_state_update(member, before, after):
    if member.bot or not after.channel:
        return

    guild_id = member.guild.id
    if not donnie_enabled.get(guild_id, False):
        return

    vc = after.channel
    if any(vc_client.channel == vc for vc_client in bot.voice_clients):
        return  # Already in the call

    try:
        vc_client = await vc.connect()
    except discord.ClientException:
        return

    audio_file = "clip1.mp3"

    try:
        while donnie_enabled.get(guild_id, False) and len(vc.members) > 1:
            vc_client.play(discord.FFmpegPCMAudio(audio_file))
            while vc_client.is_playing():
                await asyncio.sleep(1)
            await asyncio.sleep(1.5)
    finally:
        await vc_client.disconnect()

bot.run(DCTOKEN)
